http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java

```diff
@@ -1383,165 +1491,228 @@ protected void doOnError(Throwable t) {
-                            o = ((Optional) o).orElse(NOT_FOUND);
-                        }
-                        if (o == NOT_FOUND) {
-                            if (isCompletable || finalRoute.isVoid()) {
-                                singleResponse = forStatus(finalRoute.getAnnotationMetadata(), HttpStatus.OK)
-                                        .header(HttpHeaders.CONTENT_LENGTH, HttpHeaderValues.ZERO);
-                            } else {
-                                singleResponse = newNotFoundError(request);
+
+            boolean isSuspended = finalRoute.isSuspended();
+
+            Object body = finalRoute.execute();
+            if (body instanceof Optional) {
+                body = ((Optional<?>) body).orElse(null);
+            }
+
+            HttpRequest<?> incomingRequest = requestReference.get();
+            MutableHttpResponse<?> outgoingResponse;
+
+            if (body == null) {
+                if (finalRoute.isVoid()) {
+                    outgoingResponse = forStatus(finalRoute.getAnnotationMetadata());
+                    if (HttpMethod.permitsRequestBody(request.getMethod())) {
+                        outgoingResponse.header(HttpHeaders.CONTENT_LENGTH, HttpHeaderValues.ZERO);
+                    }
+                } else {
+                    outgoingResponse = newNotFoundError(request);
+                }
+            } else {
+                HttpStatus defaultHttpStatus = isErrorRoute ? HttpStatus.INTERNAL_SERVER_ERROR : HttpStatus.OK;
+                boolean isReactive = finalRoute.isAsyncOrReactive() || Publishers.isConvertibleToPublisher(body);
+                if (isReactive) {
+                    Class<?> bodyClass = body.getClass();
+                    boolean isSingle = isSingle(finalRoute, bodyClass);
+                    boolean isCompletable = !isSingle && finalRoute.isVoid() && Publishers.isCompletable(bodyClass);
+                    if (isSingle || isCompletable) {
+                        // full response case
+                        Publisher<Object> publisher = Publishers.convertPublisher(body, Publisher.class);
+                        Publishers.mapOrSupplyEmpty(publisher, new Publishers.MapOrSupplyEmpty<Object, MutableHttpResponse<?>>() {
+                            @Override
+                            public MutableHttpResponse<?>  map(Object o) {
+                                MutableHttpResponse<?> singleResponse;
+                                if (o instanceof Optional) {
+                                    Optional optional = (Optional) o;
+                                    if (optional.isPresent()) {
+                                        o = ((Optional<?>) o).get();
+                                    } else {
+                                        return supplyEmpty();
+                                    }
+                                }
+                                if (o instanceof HttpResponse) {
```

Comment:
Need to check if `o` instanceof `HttpStatus`

Reply:
Sound like a new feature, but it makes sense